src/object/cli_obj.c

```diff
@@ -7060,7 +7060,12 @@ dc_obj_coll_punch(tse_task_t *task, struct dc_object *obj, struct dtx_epoch *epo
 	if (rc != 0)
 		goto out;
 
+	leader = coa->coa_dct_nr;
+
 	if (auxi->io_retry) {
+		if (unlikely(spa->pa_auxi.shard >= obj->cob_shards_nr))
```

Comment:
it looks to me spa->pa_auxi.shard != obj->cob_shards_nr looks more safe, shard might be extended after rebuild as well?

Reply:
It is no matter if the object layout is extended, because under extended case, the shard with index "spa->pa_auxi.shard" will not disappear, it still can be used as the leader.

Reply:
Indeed, thanks for explain.